### PR TITLE
datasets: Handle converting `int16` audio data in `VoiceSample`.

### PR DESCRIPTION
**Pull request by shaper**
We saw `VoiceSample` failing the assert on `float32` audio data when playing around with the Gradio infer app and submitting an `mp3` file. We didn't dig deeper into Gradio (I'm sure it's possible to alter/convert there as well), but it seems potentially useful for `VoiceSample` to handle `int16` audio data on top of what it already handles.

### CODE REVIEW

File: ultravox/data/datasets.py
Comment (farzadab) [inline ultravox/data/datasets.py:165]: Nit: I vote to remove the comment. It makes sense for the PR, but I don't think it makes as much sense once the PR has been merged.

File: ultravox/data/datasets.py
Comment (farzadab) [inline ultravox/data/datasets.py:166]: I'm wondering if this fix obviates the need for these lines:
https://github.com/fixie-ai/ultravox/blob/main/ultravox/inference/infer.py#L70_L71

**Review by farzadab — APPROVED** [review]
LGTM, thanks!

File: ultravox/data/datasets.py
Comment (farzadab) [inline ultravox/data/datasets.py:166]: Seems like the issue is not limited to `int16`. We can get `int32` as well:
```suggestion
                self.audio = self.audio.astype(np.float32) / np.float32(32768.0)
            elif self.audio.dtype == np.int32:
                self.audio = self.audio.astype(np.float32) / np.float32(2147483648)

```

Comment (farzadab) [reply ultravox/data/datasets.py:166]: Of course another solution is to change Gradio input side:
`gr.Audio(label="Audio", show_download_button=True)` -> add `type="filepath` and load the audio file ourselves.

File: ultravox/data/datasets.py
Comment (juberti) [reply ultravox/data/datasets.py:165]: Thanks for explaining the rationale though, I was curious where you were hitting this.

File: ultravox/data/datasets.py
Comment (juberti) [reply ultravox/data/datasets.py:166]: Yeah, and possibly also in UltravoxProcessor too.

**Review by juberti — APPROVED** [review]
If it's easy, suggest adding a test to datasets_test.py, a la the existing `create_sample`.

File: ultravox/data/datasets.py
Comment (shaper) [reply ultravox/data/datasets.py:166]: Would you prefer I look at removing the two you mention as part of this PR, or separately? Probably separately is better, but asking while I learn your PR preferences.

Comment (juberti) [reply ultravox/data/datasets.py:166]: nah, go ahead and land this PR, definitely possible that cleaning up those other lines starts pulling on a thread...

Comment (shaper) [reply ultravox/data/datasets.py:166]: > Of course another solution is to change Gradio input side: `gr.Audio(label="Audio", show_download_button=True)` -> add `type="filepath` and load the audio file ourselves.

There seem to be enough code paths that can lead into this that it's worth adding your suggestion. It may also be worth the Gradio change but if so leaving for another time.

Comment (shaper) [reply ultravox/data/datasets.py:166]: Will look at the Infer and Processor bits where we may be able to remove things next.

**Comment by shaper** [comment]
> If it's easy, suggest adding a test to datasets_test.py, a la the existing `create_sample`.

Added tests, using some type hints in the test resulted in the need to enhance `VoiceSample.audio` type from `audio: Optional[np.ndarray] = None` to `audio: Optional[NDArray[np.float32]] = None` which I think is beneficial (and didn't create any other type issues elsewhere).

Will wait a bit to merge in case there's feedback on the tests.

**Review by juberti — APPROVED** [review]

**Comment by shaper** [comment]
@juberti need you or @farzadab to land as I don't have write access. Excitement! Thanks for the onboarding guidance.

**Comment by farzadab** [comment]
Merged and gave you write access. Thanks!

**Pull request merged**
